### PR TITLE
Add async FCFS policy support and other scheduling optimization

### LINKED ISSUE #577: Use async RPC from qmanager to resource
This was discussed before but I couldn't find an existing issue. The expectation is this will increase performance by allowing overlapping between qmanager and resource operations. https://github.com/flux-framework/flux-sched/issues/573#issuecomment-578420286

### PR DESCRIPTION
**Pull request by dongahn**
This PR adds a few scheduling optimization.

- Add delay scheduling optimization: it moves the invocation of the scheduling loop into prepare/check watcher callbacks so that rapidly submitted jobs will have a chance to be enqueued before they are considered in the loop. This allows for bulk processing;
- Add multi-match support: a match_multi stream RPC can take in multiple job requests, run a match for each job request and reply each matching result to the sender. This helps reduce the overheads associated with singleton match RPC.
- Limit annotation update only upto queue-depth;
- Add async multi-match support into the high-level resource API;
- Migrate FCFS queue policy to async execution.

Fixes #577, #575, #487

### CODE REVIEW

**Comment by dongahn** [comment]
@SteVwonder: this should be ready for your review.

**Comment by dongahn** [comment]
@SteVwonder: Sorry -- I just found one missing change. Let me add that and force a push unless you started your review.

**Comment by dongahn** [comment]
Ok. I pushed that one line change and squashed.

**Comment by dongahn** [comment]
Filed one issue against flux-core, which I found during the testing of this PR on Quartz at a larger scale than my dev environment: https://github.com/flux-framework/flux-core/issues/3625

File: qmanager/policies/queue_policy_fcfs_impl.hpp [hunk 1]
```diff
@@ -154,6 +157,13 @@ int queue_policy_fcfs_t<reapi_type>::handle_match_failure (int errcode)
                               (errcode == ENODEV)? "unsatisfiable"
                                                  : "match error");
     }
+    if (errcode == ENODATA && m_queue_depth_limit) {
+        // Because the scheduling loop is being terminated
+        // per queue_depth_limit, the queue should still be
+        // schedulable.
```
Comment (SteVwonder) [inline qmanager/policies/queue_policy_fcfs_impl.hpp:163]: I think I'm missing something pretty big here.  I'm confused by this comment and the git commit message.  My understanding is that the behavior you are trying to avoid is the intended behavior.  Once you hit the queue depth limit, the scheduling loop should stop and not try again until either a new job is submitted (and even then, it would technically need to be insert at a depth in the queue higher than the limit) or resources are freed by a currently running job.  Even before batching, my understanding is that it should never have been possible for a job deeper in the queue than the queue depth to be scheduled.

With this code as it is written, my understanding is that (assuming no new jobs are submitted and no jobs finish) on the next check/prep/idle loop, the scheduling loop will run again, but it will consider the exact same set of jobs that it just considered.  This means that the scheduler will always be active and never not be trying to schedule jobs when the number of pending jobs > queue depth. Is that correct?

**Review by SteVwonder — CHANGES_REQUESTED** [review]
Thanks @dongahn for putting this together. This will have huge performance benefits, especially for large bursts of jobs.  I left a few inline comments.  I also have one overarching comment w.r.t. the tracking of the scheduler loop being idle or active:

Currently, the setting of the scheduler loop being active is spread across three different files (reapi, fcfs_impl, and bf_impl) and it is done using two different interfaces (`set_sloop_active` and `set_sched_loop_active`).  The getting of this value is done in a fourth file (qmanager_callbacks) but only via one of the getter interfaces (`is_sched_loop_active`).  This distribution of the logic across many files (and layers of the scheduler) makes it harder to reason about the control flow and think through possible edge cases.  I commented in a few places where I think there are edge cases, but I'm not sure those are correct and I'm also not sure I've been able to enumerate all of the possible edge cases.

I think going forward, this will be tricky to maintain and easy for bugs to creep in (especially with the more exotic multi/batching that is possible with hybrid backfilling).  Fortunately, I **think** it would be possible to push most of this logic (at least the setting part) into a single layer (reapi).  If so, I think that would make reasoning about and maintaining that logic much easier.  Thoughts?

**Comment by dongahn** [comment]
> Currently, the setting of the scheduler loop being active is spread across three different files (reapi, fcfs_impl, and bf_impl) and it is done using two different interfaces (set_sloop_active and set_sched_loop_active). The getting of this value is done in a fourth file (qmanager_callbacks) but only via one of the getter interfaces (is_sched_loop_active). This distribution of the logic across many files (and layers of the scheduler) makes it harder to reason about the control flow and think through possible edge cases. I commented in a few places where I think there are edge cases, but I'm not sure those are correct and I'm also not sure I've been able to enumerate all of the possible edge cases.

> I think going forward, this will be tricky to maintain and easy for bugs to creep in (especially with the more exotic multi/batching that is possible with hybrid backfilling). Fortunately, I think it would be possible to push most (all?) of this logic into a single layer (reapi). If so, I think that would make reasoning about and maintaining that logic much easier. Thoughts?

I struggled with this myself as well. reapi isn't the right layer. But it may be possible to make `queue_policy_base_t` a subclass of the newly added the `queue_adapter_base_t` class. Let me try.

File: resource/modules/resource_match.cpp [hunk 2]
```diff
@@ -1813,6 +1818,76 @@ static void match_request_cb (flux_t *h, flux_msg_handler_t *w,
...
+    std::shared_ptr<resource_ctx_t> ctx = getctx ((flux_t *)arg);
+
+    if (flux_request_unpack (msg, NULL, "{s:s s:s}",
+                                          "cmd", &cmd,
+                                          "jobs", &jobs_str) < 0)
+        goto error;
+    if ( !(jobs = json_loads (jobs_str, 0, &err))) {
+        errno = ENOMEM;
+        goto error;
+    }
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/modules/resource_match.cpp:1840]: I think I tried `o` and that didn't work. Let me try again.

File: resource/modules/resource_match.cpp [hunk 3]
```diff
@@ -1813,6 +1818,76 @@ static void match_request_cb (flux_t *h, flux_msg_handler_t *w,
...
+{
+    size_t index;
+    json_t *value;
+    json_error_t err;
+    int saved_errno;
+    json_t *jobs = nullptr;
+    const char *cmd = nullptr;
+    const char *jobs_str = nullptr;
+    std::shared_ptr<resource_ctx_t> ctx = getctx ((flux_t *)arg);
+
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/modules/resource_match.cpp:1832]: Nice catch! Thanks.

File: qmanager/modules/qmanager_callbacks.cpp [hunk 4]
```diff
@@ -101,8 +102,10 @@ int qmanager_cb_t::post_sched_loop (flux_t *h, schedutil_t *schedutil,
             flux_log (h, LOG_DEBUG, "%s (id=%jd queue=%s)", note.c_str (),
                      static_cast<intmax_t> (job->id), queue_name.c_str ());
         }
-        for (job = queue->pending_begin (); job != nullptr;
+        for (job = queue->pending_begin ();
+             job != nullptr && qd < queue->get_queue_depth ();
              job = queue->pending_next ()) {
+             qd++;
```
[the remark this answers is not in the document]
Comment (dongahn) [reply qmanager/modules/qmanager_callbacks.cpp:108]: `qd` wasn't set initially in the for loop head though. I can move that to the header as well for consistency.

File: resource/hlapi/bindings/c++/reapi.hpp [hunk 5]
```diff
@@ -39,6 +39,60 @@ namespace Flux {
...
+/*! Queue adapter base API class: define a set of methods a queue
+ *  policy class (a subclass of this API class) must implement
+ *  to be able to work with reapi_t under asynchronous execution.
+ */
+class queue_adapter_base_t {
+public:
+    /*! When a match succeeds, this method is called back
+     *  by reapi_t with the matched resource information.
+     *  The implementor (e.g., queue policy class) of this method
+     *  is expected to dequeue the the job from its pending queue
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/hlapi/bindings/c++/reapi.hpp:51]: Good catch!

File: resource/hlapi/bindings/c++/reapi.hpp [hunk 6]
```diff
@@ -39,6 +39,60 @@ namespace Flux {
...
+     *  The implementor of this method is expected to dequeue the
+     *  the job from the pending job queue, if appropriate, and proceed
+     *  to the next pending job or return 0 if the scheduling loop
+     *  must be terminated per its queuing policy (e.g., FCFS).
+     *
+     *  \param           errno returned from the resource match service.
+     *                       EBUSY: resource unavailable
+     *                       ENODEV: unsatisfiable jobspec
+     *                       ENODATA: no more jobspec to process
+     *                       Others: one that can raised from match_multi RPC
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/hlapi/bindings/c++/reapi.hpp:80]: Any `errno` that can come from the call chains of `run_match`.

File: qmanager/policies/queue_policy_fcfs.hpp [hunk 7]
```diff
@@ -32,18 +32,27 @@ namespace queue_manager {
...
     virtual ~queue_policy_fcfs_t ();
     virtual int run_sched_loop (void *h, bool use_alloced_queue);
     virtual int reconstruct_resource (void *h, std::shared_ptr<job_t> job,
                                       std::string &R_out);
     virtual int apply_params ();
+    virtual int handle_match_success (int64_t jobid, const char *status,
+                                      const char *R, int64_t at, double ov);
+    virtual int handle_match_failure (int errcode);
+    virtual bool get_sloop_active ();
+    virtual void set_sloop_active (bool active);
```
[the remark this answers is not in the document]
Comment (dongahn) [reply qmanager/policies/queue_policy_fcfs.hpp:48]: I struggled with this because this class inherits from two abstract class: `queue_base_t` and the new `queue_adapter_base_t`, which is a way to adapt the abstract queue to the `reapi` without exposing too details of queue at all. I will see if this can be handled by the bottom most base queue class.

File: qmanager/policies/queue_policy_fcfs_impl.hpp [hunk 8]
```diff
@@ -52,50 +52,133 @@ int queue_policy_fcfs_t<reapi_type>::cancel_completed_jobs (void *h)
...
+
+template<class reapi_type>
+int queue_policy_fcfs_t<reapi_type>::handle_match_success (
+                                         int64_t jobid, const char *status,
+                                         const char *R, int64_t at, double ov)
+{
+    if (!is_sched_loop_active ()) {
+        errno = EINVAL;
+        return -1;
+    }
```
[the remark this answers is not in the document]
Comment (dongahn) [reply qmanager/policies/queue_policy_fcfs_impl.hpp:134]: I think the thought was this method could be called from another continuation function. Since this is a implementation of the "public" class method from `queue_adapter_base_t`, I do think it makes sense to do error checking at this level for future use.

File: qmanager/policies/queue_policy_fcfs_impl.hpp (hunk 1 above)
Comment (dongahn) [reply qmanager/policies/queue_policy_fcfs_impl.hpp:163]: Yeah I had to think about this myself as well. `queue-depth` is an optimization so that each schedule loop invocation will consider pending jobs only up to the queue depth. This way, the scheduler will not block too long going through all pending jobs in an unbounded manner. 

Now, if a loop invocation terminated with "no later jobs can be scheduled", yes the queue should become no longer schedulable and we should not invoke the loop again and again going over the exact sequence of pending jobs until the next job or resource event arises. I believe the code handles this correctly.

However, if the loop terminated with the queue-depth limit hit, the loop should still remain schedulable and the next time the loop is invoked, it should consider the next set of jobs in order to avoid starvation under our new `delay scheduling`

Imagine an extreme case where all M jobs are enqueued before we see the first invocation of the schedule loop, which terminated scheduling only queue-depth number of pending jobs (where M >> queue-depth). Because there will be no further "job" event (which we typically see from "non delay scheduling"), the schedule loop will not be invoked again until the first set of jobs finish, which can be a long time. This is a starvation that we probably don't want to see.

File: resource/modules/resource_match.cpp [hunk 9]
```diff
@@ -1813,6 +1818,76 @@ static void match_request_cb (flux_t *h, flux_msg_handler_t *w,
...
+            errno = EINVAL;
+            flux_log_error (h, "%s: existent job (%jd).",
+                            __FUNCTION__, static_cast<intmax_t> (jobid));
+            goto error;
+        }
+        if (run_match (ctx, jobid, cmd, js_str, &now, &at, &ov, R) < 0) {
+            if (errno != EBUSY && errno != ENODEV)
+                flux_log_error (ctx->h,
+                        "%s: match failed due to match error (id=%jd)",
+                        __FUNCTION__, static_cast<intmax_t> (jobid));
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/modules/resource_match.cpp:1865]: Good idea. W/ the current queuing policy, this info is still available at the queue policy level. But including such debug info more explicit makes sense to me.

File: qmanager/policies/queue_policy_fcfs_impl.hpp [hunk 10]
```diff
@@ -52,50 +52,133 @@ int queue_policy_fcfs_t<reapi_type>::cancel_completed_jobs (void *h)
...
+                      m_pending_provisional.end ());
+    m_pending_provisional.clear ();
+    m_iter = m_pending.begin ();
+    if (m_pending.empty ())
+        return 0;
+    if (!(jobs = json_array ())) {
+        errno = ENOMEM;
+        return -1;
+    }
+    if (pack_jobs (jobs) < 0)
```
[the remark this answers is not in the document]
Comment (dongahn) [reply qmanager/policies/queue_policy_fcfs_impl.hpp:107]: Yes we are on the same page. For now, we should control this behavior with queue-depth parameter. How about changing the limit to 32 as part of this PR?

[the remark this answers is not in the document]
Comment (dongahn) [reply qmanager/policies/queue_policy_fcfs_impl.hpp:107]: https://github.com/flux-framework/flux-sched/issues/830 opened to track this.

File: qmanager/policies/base/queue_policy_base_impl.hpp [hunk 11]
```diff
@@ -310,9 +347,17 @@ int queue_policy_base_impl_t::remove (flux_jobid_t id)
...
-        m_pending.erase ({static_cast<double> (job->priority),
-                          static_cast<double> (job->t_submit),
-                          static_cast<double> (job->t_stamps.pending_ts)});
+        sz = m_pending.erase (
+                 {static_cast<double> (job->priority),
+                  static_cast<double> (job->t_submit),
+                  static_cast<double> (job->t_stamps.pending_ts)});
+        if (sz == 0) {
+            // job must be in m_pending_provisional in this case
+            m_pending_provisional.erase (
```
[the remark this answers is not in the document]
Comment (dongahn) [reply qmanager/policies/base/queue_policy_base_impl.hpp:356]: Agreed. I will add a conditional statement with `errno=ENOENT` in this case. But won't log since that should be handled by the upper layer.

File: qmanager/policies/base/queue_policy_base_impl.hpp [hunk 12]
```diff
@@ -464,18 +561,42 @@ int queue_policy_base_impl_t::pending_reprioritize (flux_jobid_t id,
     m_pending.erase ({static_cast<double> (job->priority),
                       static_cast<double> (job->t_submit),
                       static_cast<double> (job->t_stamps.pending_ts)});
+    sz = m_pending.erase (
```
[the remark this answers is not in the document]
Comment (dongahn) [reply qmanager/policies/base/queue_policy_base_impl.hpp:564]: Wow. This is actually a merge error from git rebase... Will fix!

File: qmanager/policies/base/queue_policy_base_impl.hpp [hunk 13]
```diff
@@ -464,18 +561,42 @@ int queue_policy_base_impl_t::pending_reprioritize (flux_jobid_t id,
...
+        auto res = m_pending.insert (std::pair<std::vector<double>,
+                                               flux_jobid_t> (
+                       {static_cast<double> (job->priority),
+                        static_cast<double> (job->t_submit),
+                        static_cast<double> (job->t_stamps.pending_ts)}, job->id));
+        if (!res.second) {
+            errno = ENOMEM;
+            return -1;
+        }
+    } else {
```
[the remark this answers is not in the document]
Comment (dongahn) [reply qmanager/policies/base/queue_policy_base_impl.hpp:587]: @SteVwonder: thank you for catching this! yes this is a bug and crept up during that merge issue as well.

File: qmanager/policies/queue_policy_fcfs_impl.hpp (hunk 1 above)
Comment (dongahn) [reply qmanager/policies/queue_policy_fcfs_impl.hpp:163]: I will mark it resolved. If you disagree, let me know.

Comment (SteVwonder) [reply qmanager/policies/queue_policy_fcfs_impl.hpp:163]: AH! That makes sense.  Thanks for taking the time to explain.  Good catch!

File: resource/modules/resource_match.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/modules/resource_match.cpp:1846]: `O` doesn't seem to work either.

**Comment by dongahn** [comment]
@SteVwonder: I believe I have addressed all of your requests. Please take a quick look at the new commits. Once you are okay, I will squash the intermediate commits.

**Review by SteVwonder — APPROVED** [review]
LGTM! Thanks @dongahn
